src/main/scala/org/constellation/p2p/DataResolver.scala

```diff
@@ -39,57 +39,58 @@ class DataResolver extends StrictLogging {
+    priorityClient: IO[Option[PeerApiClient]] = IO.pure(None)
+  )(implicit apiTimeout: Duration = 3.seconds, m: Manifest[T], dao: DAO): List[IO[T]] =
     hashes.map { hash =>
       val dataHash = BigInt(hash.getBytes)
 
-      resolveData[T](hash, endpoint, priorityClient.toSeq ++ pool.toSeq.sortBy { p =>
-        BigInt(p.id.hex.getBytes()) ^ dataHash
-      }, store)
+      val sorted = priorityClient.map { client =>
+        pool.map { peers =>
+          client.toList ++ peers.sortBy { p =>
+            BigInt(p.id.hex.getBytes()) ^ dataHash
+          }
+        }
+      }.flatten
+
+      resolveData[T](hash, endpoint, sorted , store)
     }
 
   def resolveData[T <: AnyRef](
     hash: String,
     endpoint: String,
-    sortedPeers: Iterable[PeerApiClient],
+    sortedPeers: IO[List[PeerApiClient]],
     store: T => Any,
     maxErrors: Int = 10
-  )(implicit apiTimeout: Duration = 3.seconds, m: Manifest[T], dao: DAO): IO[Option[T]] = {
-    val storeIO = (t: T) => IO { store(t) }.void
+  )(implicit apiTimeout: Duration = 3.seconds, m: Manifest[T], dao: DAO): IO[T] = {
 
-    def makeAttempt(innerPeers: Iterable[PeerApiClient], errorsSoFar: Int = 0): IO[Option[T]] =
+    def makeAttempt(innerPeers: Iterable[PeerApiClient], allPeers: Iterable[PeerApiClient], errorsSoFar: Int = 0): IO[T] =
       innerPeers match {
         case _ if errorsSoFar >= maxErrors =>
           IO.raiseError(DataResolutionMaxErrors(endpoint, hash))
         case Nil =>
           IO.raiseError(
-            DataResolutionOutOfPeers(dao.id.short, endpoint, hash, sortedPeers.map(_.id.short))
+            DataResolutionOutOfPeers(dao.id.short, endpoint, hash, allPeers.map(_.id.short))
           )
         case head :: tail =>
-          getData[T](hash, endpoint, head, store).handleErrorWith {
+          getData[T](hash, endpoint, head, store).flatMap {
+            case Some(a) => IO.pure(a)
+            case None    => makeAttempt(tail, allPeers, errorsSoFar + 1)
+          }.handleErrorWith {
             case e if tail.isEmpty => IO.raiseError(e)
             case e =>
               logger.error(
                 s"Failed to resolve with host=${head.client.hostPortForLogging}, trying next peer",
                 e
               )
-              makeAttempt(tail, errorsSoFar + 1)
-          }.flatTap(_.map(storeIO).getOrElse(IO.unit))
-            .flatMap { response =>
-              if (response.isEmpty) {
-                logger.warn(
-                  s"Empty response resolving with host=${head.client.hostPortForLogging}, trying next peer"
-                )
-                makeAttempt(tail, errorsSoFar + 1)
-              } else {
-                IO.pure(response)
-              }
-            }
+              makeAttempt(tail, allPeers, errorsSoFar + 1)
+          }
+
       }
-    makeAttempt(sortedPeers)
+    sortedPeers.flatMap(peers => makeAttempt(peers, peers).flatMap { t =>
+      IO.pure(store(t)) *> IO.pure(t)
```

Comment:
`store(t)` should be wrapped in `pure`. it should use `.delay` or just apply method:
`IO(store(t)) *> IO.pure(t)`
but if you want to just make a side effect like this, it should be:
`.flatTap(t => IO(store(t))`

Reply:
Good point we shouldn't use pure for side effects.